Scrubble/ps3.py

```diff
@@ -0,0 +1,398 @@
+
+    * The hand is displayed.
+    
+    * The user may input a word.
+
+    * When any word is entered (valid or invalid), it uses up letters
+      from the hand.
+
+    * An invalid word is rejected, and a message is displayed asking
+      the user to choose another word.
+
+    * After every valid word: the score for that word is displayed,
+      the remaining letters in the hand are displayed, and the user
+      is asked to input another word.
+
+    * The sum of the word scores is displayed when the hand finishes.
+
+    * The hand finishes when there are no more unused letters.
+      The user can also finish playing the hand by inputing two 
+      exclamation points (the string '!!') instead of a word.
+
+      hand: dictionary (string -> int)
+      word_list: list of lowercase strings
+      returns: the total score for the hand
+      
+    """
+
+    score = 0
+    while len(hand) != 0:
+        print()
+        print(f"Current Hand:", end=' ')
+        print(display_hand(hand))
+        users_word = input("Enter word, or “!!” to indicate that you are finished: ")
+        if users_word == "!!":
+            break
+
+        else:
+            if is_valid_word(users_word, hand, word_list):
+                temp_score = get_word_score(users_word, calculate_handlen(hand))
+                print(f'"{users_word}" earned {temp_score} points.', end='')
+                score += temp_score
+                print(f'Total: {score} points')
+            else:
+                print('This is not a valid word. Please choose another word.')
+            hand = update_hand(hand, users_word)
+    else:
+        print()
+        print('Ran out of letters')
+
+    print('Total score for this hand: ', score)
+    return score
+
+
+# Problem #6: Playing a game
+def substitute_hand(hand, letter):
+    """ 
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+
+    For example:
+        substitute_hand({'h':1, 'e':1, 'l':2, 'o':1}, 'l')
+    might return:
+        {'h':1, 'e':1, 'o':1, 'x':2} -> if the new letter is 'x'
+    The new letter should not be 'h', 'e', 'l', or 'o' since those letters were
+    already in the hand.
+    
+    hand: dictionary (string -> int)
+    letter: string
+    returns: dictionary (string -> int)
+    """
+
+    all_letters = VOWELS.union(CONSONANTS)
+    new_hand = hand.copy()
+
+    if letter in new_hand.keys():
+
+        # just take set with all letters without choosen letter, make new list and replace keys
+        new_hand[random.choice(list(all_letters.difference(set(letter))))] = new_hand.pop(letter)
+
+    return new_hand
```

Comment:
Check if this code works according to the docstring. Your code can change a letter on the letter that's already in hand. Sometimes, it's even returns a hand without changed letter(only 6 letters).